New cookies recipe (previous one was a bit silly)

UmeFuu opened the pull request:
**What does this PR do:**
The previous cookie recipe was 1 dough + 1 chocolate bar, throw that in the oven and voilá!
This one adds a new dough specially made for cookies and pastry (yum) and some more simple steps before the oven.

Chocolate crumbles/pile of chocolate: Just cut a chocolate bar into pieces.
Create the dough: mix 10u of milk, 10u of flour and 5u of sugar. You'll get pastry dough!
Flatten the dough using the rolling pin (duh) and then cut perfect circles with the new circular cutter™.
Add the unbaked cookies to the oven, and you'll get six perfect sugar cookies.
Optionally, if you add a chocolate pile to the unbaked cookies before putting them into the oven, you'll get six chocolate chips cookies! (Just the regular cookies)

**I illustrated this really complex process just for you:**
![image](https://user-images.githubusercontent.com/35434208/52684646-b2b4ce80-2f25-11e9-934a-75f126f8f670.png)
---------------------
Edit: Now with (optional) choco chips 
![image](https://user-images.githubusercontent.com/35434208/52742829-67e39700-2fb7-11e9-9d53-83d1b6025cc1.png)


**Changelog:**
:cl:
add: New recipe for cookies and sugar cookies, new ingredients such as pastry dough, flat pastry dough, unbaked cookies and a new tool called circular cutter (Added to the dinnerware vendor). New chem mix to create a specific type of dough for cookies and pastries.

del: Removed old cookie/sugarcookie recipe

tweak: added reagents of the cookies (They now actually taste like sugar and chocolate):
- sugar = 3
- hot_coco = 5

Also adjusted the sugar cookies nutriment from 3 to 1, just like chocolate cookies (They're made of the same dough after all)
Chocolate bar can now be cutted into pieces/chocolate pile

imageadd: added cookie(pastry) dough, flat cookie(pastry) dough, unbaked cookies/choco and circular cutter sprites. 
(Disclaimer: The cookie dough and flat cookie dough are edits of the normal doughs, pretty obvious but credits to who sprited them. Same with the tray with the unbaked cookies)
/:cl:

Raze500 commented:
Flawless Changelog, +10 In Changeling meat.

taukausanake commented:
Awww, I was hoping for chocolate chips or something. Am happy with what the recipe is now though

UmeFuu commented:
> Awww, I was hoping for chocolate chips or something. Am happy with what the recipe is now though

I could add that in the future, as I'm planning on expanding this base recipe to make a lot of different cookies ^^

AzuleUtama commented:
Now this is a tasty PR right here.

Citinited commented on code/game/objects/items/weapons/storage/bags.dm:
```diff
@@ -460,6 +460,20 @@
+		handle_item_insertion(C)    // Done this way so the tray actually has the cookies visible when spawned
+	rebuild_overlays()
+
+/obj/item/storage/bag/tray/sugarcookies_tray/New() /// By Azule Utama, thank you a lot!
```
I'd suggest making this a subtype of the cookie tray, then adding a var to the tray that tells you what kind of cookie is meant to be created on the tray, then just spawning that var in the for loop. It'll save you copy-pasting code, and is also more OOP.

UmeFuu replied:
Sounds like a good idea, I'm kinda new to all this coding thing, so I just read what's already done and try to make something out of it. So I can try.

Citinited replied:
Aye, no worries. It's all part of the learning experience - a good rule of thumb is that if you end up with the exact same code in two different places you can probably make a proc that applies to both cases. Good luck!

UmeFuu replied:
I did it, so it's fixed. I'm doing some final tests to all the suggestions you made and I'll change the files again.

Citinited commented on code/modules/food_and_drinks/food/foods/ingredients.dm:
```diff
@@ -129,6 +129,50 @@
+/obj/item/reagent_containers/food/snacks/cookiedough/attackby(obj/item/I, mob/user, params)
+	if(istype(I, /obj/item/kitchen/rollingpin))
+		if(isturf(loc))
+			new /obj/item/reagent_containers/food/snacks/cookiedough_flat(loc)
```
Instead of doing it this way, why not have a `flat` var that determines how it acts? This saves you having to make a subtype and duplicating code, and is more object-oriented (and keeps object deletion and creation down which is a benefit I guess)

e.g. (psuedocode below)

```
/obj/item/reagent_containers/food/snacks/cookiedough/attackby(mob/user, obj/item/W)
    if(istype(W, rollingpin) && !flat)
        to_chat(user, "you flatten [src].")
        flat = TRUE
    else if(istype(W, cookie_cutter) && flat)
        to_chat(user, "you cut [src] into cookie shapes.")
        new obj/item/food/raw_cookie(loc)
        qdel(src)
    else
        ..()
```

UmeFuu replied:
Here I just copied the normal dough system (You know, rolling pin, knife) but it's a great idea too. I'll add that code and test if it works properly (As I see it, it should), and credit you if you don't mind

Citinited replied:
I don't generally care about being credited for minor things like this, if you want to go ahead though 👍 

Also, while copy-pasting is fine for some things, I'd try to keep that to a minimum, as we have so much substandard code copy-pasting it may be detrimental to how good your overall code is. Just my two cents 😄

UmeFuu replied:
Also fixed this

Citinited commented on code/modules/reagents/chemistry/recipes/food.dm:
```diff
@@ -102,6 +102,22 @@
+
+/datum/chemical_reaction/cookiedough/on_reaction(datum/reagents/holder, created_volume)
+	var/location = get_turf(holder.my_atom)
+	for(var/i = 1, i <= created_volume, i++)
```
Should do `for(var/i in 1 to created_volume)` here.

UmeFuu replied:
I'm not sure why or what does that do, but it works as it is. It's the same code as the normal dough.

Citinited replied:
It works the same way, but the way I've put is how we generally format for loops in paradise. I think it may also be faster?

Citinited commented on code/modules/reagents/chemistry/recipes/food.dm:
```diff
@@ -102,6 +102,22 @@
+	mix_message = "The ingredients form a dough. It smells sweet and yummy."
+
+/datum/chemical_reaction/cookiedough/on_reaction(datum/reagents/holder, created_volume)
+	var/location = get_turf(holder.my_atom)
```
Maybe should use `drop_location()` here. Otherwise people could become cookie dough factories by eating the constituent ingredients and have a bunch of dough pop out of them 😄

UmeFuu replied:
_That's actually kinda intended so I wouldn't fix it, I just find it hilarious_

Citinited replied:
🤔

Mitchs98 replied:
People can do this with normal dough, I'd leave it in for comedic value myself. :man_shrugging:

UmeFuu replied:
I actually do it all the time in front of people when I'm the chef

Superhats commented:
I only agree with this if we can add raisins too

FalseIncarnate commented:
Could make fortune cookies use this as well, not a bad addition though.

I’d like to suggest that the sugar cookie recipe takes some sugar in the oven, rather than just being the dough. Would help keep recipes distinct to hopefully avoid potential conflicts in the future

UmeFuu commented:
> I only agree with this if we can add raisins too

It could be, I would add that in the future just to make everyone mad (People hate raisins for some reason)



> Could make fortune cookies use this as well, not a bad addition though.
> 
> I’d like to suggest that the sugar cookie recipe takes some sugar in the oven, rather than just being the dough. Would help keep recipes distinct to hopefully avoid potential conflicts in the future

Fortune cookies are made in a special way because they use specific papers to go into them, and I wouldn't like to mess with that for now. Their recipe is already as complex as this one, they're just made of normal dough because they're not sweet.
For the sugar cookies, the dough already has sugar in it, so I find it kinda unnecessary to add even more sugar to the oven

Bookdrizzt commented:
> It could be, I would add that in the future just to make everyone mad (People hate raisins for some reason)
I guess you could say that it's for no raisin.

Also, good to see this! I can't wait to see some people baking cookies and handing them out, or cookie shops popping up just like when the abandoned pizza place was added.

UmeFuu commented:
@taukausanake you asked it, you got it. You can add chocolate chips/crumbles/whatever you want to call them to the cookies before the oven, so there's no need to put a chocolate bar into the oven now.
:3 
![image](https://user-images.githubusercontent.com/35434208/52742778-4682ab00-2fb7-11e9-87f8-25a26a0c144c.png)

Fox-McCloud commented:
Adding even more one to two use kitchen tools isn't desirable. Just use the knife.

Fox-McCloud commented on code/modules/food_and_drinks/food/foods/ingredients.dm:
```diff
@@ -140,6 +205,19 @@
+		else
+			to_chat(user, "<span class='notice'>You need to put [src] on a surface to cut it out!</span>")
+	else
+		..()
```
return the value.

UmeFuu replied:
Could you be a little more specific? I don't know where I should put the return, as it wasn't in the example I used to base my code (The original dough and rolling pin code), I looked at it because maybe I missed it but isn't there neither

UmeFuu replied:
@Fox-McCloud  It works perfectly as it is so It'd be good if you tell me if it's necessary to add it, why and how
I'm scared to do it wrong and break something

Fox-McCloud commented on code/modules/food_and_drinks/food/foods/ingredients.dm:
```diff
@@ -129,6 +129,71 @@
+		else
+			to_chat(user, "<span class='notice'>You need to put [src] on a surface to add this</span>")
+	else
+		..()
```
same

Fox-McCloud commented on code/modules/food_and_drinks/food/foods/ingredients.dm:
```diff
@@ -129,6 +129,71 @@
+		..()
+
+
+/obj/item/reagent_containers/food/snacks/rawcookies/
```
no `/` on the end.

Fox-McCloud commented on code/game/objects/items/weapons/storage/bags.dm:
```diff
@@ -460,6 +460,28 @@
 	return ..()
 
 
+/obj/item/storage/bag/tray/cookies_tray/
```
no `/`

UmeFuu commented:
> Adding even more one to two use kitchen tools isn't desirable. Just use the knife.

I think the cutter will be necessary, as I mentioned before and people is excited about it, this is a base recipe for more variations. Just using the knife isn't interesting/accurate enough, or even possible, as one of the future recipes I'm planning will be just about normal flat dough + cutter, and the knife already cuts that into dough slices.

>Differentiating between raw cookie dough and dough is kinda gross

I think I don't get what you meant by that

FalseIncarnate commented:
My comment about fortune cookies was meant as a “use the cookie dough instead of normal dough”.

My comment about adjusting the sugar cookie recipe to be more than just cookie dough is to avoid any issues with recipe selection. It shouldn’t be an issue, but I am aware there at least was a time where a recipe had to be listed before another recipe in the code or else it would make the wrong one because it would use the less complex one that was listed first. I’m not sure if that is still an issue, but definitely worth checking and ensuring isn’t an issue moving forward with this.

Because the kitchen machines aren’t stasis reagent containers, they will provide reactions on chemicals added, so we want to make sure we aren’t interfering with existing recipes when new chem reactions are added (saline glucose, for example, required altering the taffy recipe because the ingredients happened to qualify for the reaction)

Fox-McCloud commented:
>I think I don't get what you meant by that,

It meants adding additional items, steps, and tools *purely* because "realism" isn't inherently a good thing, from a design perspective. Adding new tools or new variants should be what you do when you have a mechanical reason to do so.

UmeFuu commented:
> My comment about fortune cookies was meant as a “use the cookie dough instead of normal dough”.
> 
> My comment about adjusting the sugar cookie recipe to be more than just cookie dough is to avoid any issues with recipe selection. It shouldn’t be an issue, but I am aware there at least was a time where a recipe had to be listed before another recipe in the code or else it would make the wrong one because it would use the less complex one that was listed first. I’m not sure if that is still an issue, but definitely worth checking and ensuring isn’t an issue moving forward with this.
> 
> Because the kitchen machines aren’t stasis reagent containers, they will provide reactions on chemicals added, so we want to make sure we aren’t interfering with existing recipes when new chem reactions are added (saline glucose, for example, required altering the taffy recipe because the ingredients happened to qualify for the reaction)

You are right about the ussue, but If you look at the recipe, I'm driving it to be all outside the oven, and using it at the end to just bake the final product. Everything else happens outside the oven. That's what I'm planning for the rest of recipes too.

As I said, I don't see the need to change the fortune cookies, as they seem pretty accurate with non sweet pastry dough (I'll change the name of the cookie dough to that), just normal dough

UmeFuu commented on code/modules/reagents/chemistry/recipes/food.dm:
```diff
@@ -102,6 +102,22 @@
+	name = "Dough"
+	id = "dough"
+	result = null
+	required_reagents = list("milk" = 10, "flour" = 10, "sugar" = 5)
```
You're right, I tried to overlook if there were more recipes that used all that, but I missed that one. Good thing is I was planning on adding 5u of egg to the reaction, and I think only one or two recipes uses it as a reagent and not an item. I'll fix that.

UmeFuu replied:
"Bake 10u Flour, 1 Egg, 5u Sugar, 5u Milk, 1 Golden Apple" the quantities are below the pastry dough reaction, and the apple tart as a baked recipe, none of them interfere with the other. I was thinking about adding the 5u egg, but if I want to make a chocolate flavored pastry dough, it would interfere with the chocolate pudding recipe. I'll think about it later on.

UmeFuu commented:
I renamed the cookie dough (and flat version), now it's pastry dough. I just changed the name of the object for now and not the path as I see it unnecessary to change and re-upload every file of the recipe just for that. I'll fix it when I add more variations.

Mitchs98 commented:
I see no issue with adding in cookie cutters. It adds RP factor. Not just for 'realism'. Which, in a text based game largely helped by items, is a good thing. Same discussion as when I made my sushi mat...not gonna rehash that.

Back on point. This PR sounds neat. Definitely can't wait for more cookies and cookie shapes later on the line. :+1:

Fox-McCloud commented:
RP factor and realism are the same thing, at the core.

We don't base design around them; adding additional steps and tools purely for the sake of it isn't desirable, nor is it good design. If an existing tool or method already exists that is tentatively believable, that's good enough.

If new steps are *required* then something that fits from an RP or realistic standpoint should be introduced, of course, but if there's no point or purpose other than *purely* RP/realism, then it's a bad design.

If something exists purely for RP (aesthetics, to a lesser degree bar drinks), then it existing on its own merit is fine, but that's different. That's not a mechanic, system, or process the player engages with; that's an end product.

UmeFuu commented:
It's a cutter, for more upcoming recipes, not only this one
It's a necessary tool for my project and people seems to like it and don't seem bothered by the addition of a new tool
I don't see the issue with it

DrawitRalph commented:
I don't see how it's a bad design. It opens de door for more complex recipes in the future, which I'm sure any chef main would appreciate. Besides, the sprite is just too cute to let it go to waste.

Ty-Omaha commented:
My brain activated freshly baked cookie smell when I saw this

FalseIncarnate commented:
I apparently had been misreading things and thinking you just slapped the pastry dough in the oven to get cookies and had entirely glossed over the whole bit with trays and cutters. That would explain my concern about recipe conflicts with different cookies, and entirely sidesteps that concern. My bad, I should probably read stuff more carefully.

Still slightly concerned there might be an existing recipe that would accidentally cause pastry dough reaction, but I’m probably wrong. I’ll dig through recipes and post again if I find one. In the meantime, my concerns are resolved for the current version of this PR

Raze500 commented:
This will probably will never get merged due the fun fact.

variableundefined commented on code/game/objects/items/weapons/storage/bags.dm:
```diff
@@ -460,6 +460,28 @@
+		var/obj/item/C = new cookie(src)
+		handle_item_insertion(C)    // Done this way so the tray actually has the cookies visible when spawned
+	rebuild_overlays()
+/*
```
No leftover code please.

Ume-D replied:
Oops, I had some commented code while testing ad I forgot to delete it later on. Gonna fix that

variableundefined commented on code/game/objects/items/weapons/storage/bags.dm:
```diff
@@ -460,6 +460,28 @@
+/obj/item/storage/bag/tray/cookies_tray/sugarcookie
+	cookie = /obj/item/reagent_containers/food/snacks/sugarcookie
+
+//obj/item/storage/bag/tray/cookies_tray/cookie
```
Same

variableundefined commented on code/modules/food_and_drinks/food/foods/ingredients.dm:
```diff
@@ -129,6 +129,71 @@
+			update_icon()
+		else
+			to_chat(user, "<span class='notice'>You need to put [src] on a surface to roll it out!</span>")
+	else if (istype(I, /obj/item/kitchen/cutter) && flat)
```
No spacing between if and ( please

variableundefined commented on code/modules/food_and_drinks/food/foods/ingredients.dm:
```diff
@@ -129,6 +129,71 @@
+		else
+			to_chat(user, "<span class='notice'>You need to put [src] on a surface to cut it out!</span>")
+	else
+		return..()
```
Spacing between return and ..()

variableundefined commented on code/modules/food_and_drinks/food/foods/ingredients.dm:
```diff
@@ -129,6 +129,71 @@
+		else
+			to_chat(user, "<span class='notice'>You need to put [src] on a surface to add this</span>")
+	else
+		return..()
```
Spacing

variableundefined commented on code/modules/food_and_drinks/food/foods/ingredients.dm:
```diff
@@ -140,6 +205,19 @@
+		else
+			to_chat(user, "<span class='notice'>You need to put [src] on a surface to cut it out!</span>")
+	else
+		return..()
```
Spacing

variableundefined commented on code/modules/food_and_drinks/recipes/recipes_oven.dm:
```diff
@@ -163,13 +163,22 @@
 	)
-	result = /obj/item/reagent_containers/food/snacks/cookie
+	result = /obj/item/storage/bag/tray/cookies_tray/sugarcookie
+///	result = /obj/item/storage/bag/tray/sugarcookies_tray
```
No leftover code

variableundefined commented on code/modules/reagents/chemistry/recipes/food.dm:
```diff
@@ -102,6 +102,22 @@
+
+/datum/chemical_reaction/cookiedough/on_reaction(datum/reagents/holder, created_volume)
+	var/location = get_turf(holder.my_atom)
+	for(var/i = 1, i <= created_volume, i++)
```
Please use the for(var/i in 1 to created_volume) - its faster and in our style guide. It is also easier to understand

Ume-D replied:
I used it as it was like this for the normal dough, should I change the normal dough too?

UmeFuu replied:
(Sorry forgot to switch accounts before replying, I'm working on two separated projects with the same code, so I had to fork something twice with different accounts)

variableundefined commented:
Try to squash your commits next time
